File: 23-Graph/python/06_detect_a_cycle_in_directed_graph_using_dfs.py
from typing import List

class Solution:
    def isCyclic(self, N: int, adj: List[List[int]]) -> bool:
        visited = [0] * N

        def dfs_check(node: int) -> bool:
            visited[node] = 1
            for neighbor in adj[node]:
                if visited[neighbor] == 1:
                    return True
                elif visited[neighbor] == 0:
                    if dfs_check(neighbor):
                        return True
            visited[node] = 2
            return False

        for i in range(N):
            if visited[i] == 0:
                if dfs_check(i):
                    return True
        return False

File: 23-Graph/cpp/06-detect-a-cycle-in-directed-graph-using-dfs.cpp
#include <vector>

using namespace std;

class Solution {
private:
    bool dfsCheck(int node, vector<vector<int>>& adj, vector<int>& visited) {
        visited[node] = 1;
        for (int neighbor : adj[node]) {
            if (visited[neighbor] == 1) {
                return true;
            } else if (visited[neighbor] == 0) {
                if (dfsCheck(neighbor, adj, visited)) {
                    return true;
                }
            }
        }
        visited[node] = 2;
        return false;
    }

public:
    bool isCyclic(int N, vector<vector<int>>& adj) {
        vector<int> visited(N, 0);
        for (int i = 0; i < N; i++) {
            if (visited[i] == 0) {
                if (dfsCheck(i, adj, visited)) {
                    return true;
                }
            }
        }
        return false;
    }
};
